ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,18 +702,91 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
+    StoreFindToken expectedEndToken = new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId, incarnationId);
     expectedEndToken.setBytesRead(bytesRead);
+    expectedEndToken.setInclusive();
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.EMPTY_SET, expectedEndToken);
 
     // create a token that is not past the index end offset on startup after recovery. Should work as expected
-    startToken = new StoreFindToken(lastRecordOffset, oldSessionId);
-    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId);
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, incarnationId);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.singleton(newId), expectedEndToken);
+
+    // same test as above, but with old incarnationId
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on startup after recovery with old incarnationId
+    startToken = new StoreFindToken(secondRecordFileSpan.getEndOffset(), oldSessionId, oldIncarnationId);
+    // token should get reset internally, returning all the keys and the returned token should be correct (offset in
+    // it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // create a token that is not past the index end offset on startup after recovery, but with old incarnationId.
+    // again the token should be reset internally, all keys should be returned and the returned token should be correct
+    // (offset in it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, oldIncarnationId);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // test cases where no new puts/deletes have gone in after restart.
+    // after restart, replica 1(assuming no new puts have gone in) will return a journal based token referring to
+    // logEndOffsetOnStartUp to replica2.
+    // But on the subsequent findEntriesSince() call, this token is considered to be non-inclusive and hence
+    // replica1 will start returning entries from the next blob which resides after logEndOffSetOnStartup.
+    // In other words, one blob will be missed out in findEntriesSince()
```

Comment:
well - the test should say that "no entries are missed", since that is the good case. In other words, the test should not be for the existence of the bug, but the absence of it.